File: API_scripts/custom_vision.js
// load env var
require("dotenv").config({ path: "./../.env" });

// util module
const util = require("util");

// Filesystem module
const fs = require("fs");

//* load env vars ---------------------------------------------------------------------------------

// Custom vision key
const training_key = process.env.training_key;

// Custom vision key
const prediction_key = process.env.prediction_key;

// Custom vision key
const prediction_resource_id = process.env.prediction_resource_id;

/*CUSTOM VISION API**********************************************************************************************/

// Custom vision endpoint
custom_vision_endpoint = "https://eastus.api.cognitive.microsoft.com/";

const TrainingApi = require("@azure/cognitiveservices-customvision-training");
const PredictionApi = require("@azure/cognitiveservices-customvision-prediction");
const msRest = require("@azure/ms-rest-js");
const { resolveNaptr } = require("dns");

const setTimeoutPromise = util.promisify(setTimeout);

// Keys
const trainingKey = training_key;
const predictionKey = prediction_key;
const predictionResourceId = prediction_resource_id;

// Root where images are stored
const sampleDataRoot = "path to your data file";

// Custom vision endpoint
const endPoint = custom_vision_endpoint;

const credentials = new msRest.ApiKeyCredentials({
  inHeader: { "Training-key": trainingKey }
});
const trainer = new TrainingApi.TrainingAPIClient(credentials, endPoint);

// Prediction model **********************************************************************************
const publishIterationName = "classifyModel";

try {
  (async () => {
    console.log("Creating project...");

    // Set project name
    const sampleProject = await trainer.createProject("Project Custom Vision");

    console.log("Sample project ID: ", sampleProject.id);

    module.exports = { sampleProject };

    // Create tag names
    const cruiseTag = await trainer.createTag(sampleProject.id, "cruise");
    const otherTag = await trainer.createTag(sampleProject.id, "other");

    console.log("Adding images...");
    let fileUploadPromises = [];

    const cruiseDir = `${sampleDataRoot}/cruise`;
    const cruiseFiles = fs.readdirSync(cruiseDir);
    cruiseFiles.forEach(file => {
      fileUploadPromises.push(
        trainer.createImagesFromData(
          sampleProject.id,
          fs.readFileSync(`${cruiseDir}/${file}`),
          { tagIds: [cruiseTag.id] }
        )
      );
    });

    const otherDir = `${sampleDataRoot}/other`;
    const otherFiles = fs.readdirSync(otherDir);
    otherFiles.forEach(file => {
      fileUploadPromises.push(
        trainer.createImagesFromData(
          sampleProject.id,
          fs.readFileSync(`${otherDir}/${file}`),
          { tagIds: [otherTag.id] }
        )
      );
    });

    await Promise.all(fileUploadPromises);

    console.log("Training...");
    let trainingIteration = await trainer.trainProject(sampleProject.id);

    // Wait for training to complete
    console.log("Training started...");
    while (trainingIteration.status == "Training") {
      console.log("Training status: " + trainingIteration.status);
      await setTimeoutPromise(1000, null);
      trainingIteration = await trainer.getIteration(
        sampleProject.id,
        trainingIteration.id
      );
    }
    console.log("Training status: " + trainingIteration.status);

    // Publish the iteration to the end point
    await trainer.publishIteration(
      sampleProject.id,
      trainingIteration.id,
      publishIterationName,
      predictionResourceId
    );
  })();
} catch (err) {
  console.log(`Something went wrong: ${err}`);
}

File: API_scripts/custom_vision.py
import os
import time

# Load system variables with dotenv
from dotenv import load_dotenv
load_dotenv()

# Load keys and urls

# Primary key
_key = os.environ["_key"]

#Translator key
translate_key = os.environ["translate_key"]

# Key sentiment
_sentiment_key = os.environ["sentiment_key"]

_base_url = os.environ["base_url"]

_sentiment_url = os.environ["sentiment_url"]

# Custom vision key
training_key = os.environ["training_key"]

# Custom vision key
prediction_key = os.environ["prediction_key"]

# Custom vision key
prediction_resource_id = os.environ["prediction_resource_id"]

COGSVCS_CLIENTURL = _base_url
COGSVCS_KEY = _key
COGSVCS_REGION = 'eastus'

# Create vision_client
from msrest.authentication import CognitiveServicesCredentials
from azure.cognitiveservices.vision.computervision import ComputerVisionClient
from azure.cognitiveservices.vision.computervision.models import ComputerVisionErrorException


# Create the Custom Vision project
from azure.cognitiveservices.vision.customvision.training import CustomVisionTrainingClient
from azure.cognitiveservices.vision.customvision.training.models import ImageFileCreateBatch, ImageFileCreateEntry
from azure.cognitiveservices.vision.customvision.prediction import CustomVisionPredictionClient
from msrest.authentication import ApiKeyCredentials

# Custom vision endpoint
custom_vision_endpoint = "https://eastus.api.cognitive.microsoft.com/"


vision_credentials = CognitiveServicesCredentials(COGSVCS_KEY)
vision_client = ComputerVisionClient(COGSVCS_CLIENTURL, vision_credentials)

person_group_id = 'reactor'


publish_iteration_name = "classifyModel"

credentials = ApiKeyCredentials(in_headers={"Training-key": training_key})
trainer = CustomVisionTrainingClient(custom_vision_endpoint, credentials)

# Create a new project
print ("Creating project...")
project = trainer.create_project("CV Project")

# Make two tags in the new project
cruise = trainer.create_tag(project.id, "cruise")
other = trainer.create_tag(project.id, "other")

# Add the base path to the downloaded image files
base_image_url = ""

print("Adding images...")

# Image list to save images
image_list = []

 # Add path in from the base_image_url where you have the cruise ship images folder
directory = os.fsencode(base_image_url+"cruise/")

# Loop through image file folder
for file in os.listdir(directory):

    file_name = os.fsdecode(file)
    print("FILE NAME cruise: ", file_name)
    with open(base_image_url + "cruise/" + file_name, "rb") as image_contents:
        image_list.append(ImageFileCreateEntry(name=file_name, contents=image_contents.read(), tag_ids=[cruise.id]))


# Add path in from the base_image_url where you have the other images folder
directory = os.fsencode(base_image_url+"other/")


# Loop through image file folder
for file in os.listdir(directory):
    file_name = os.fsdecode(file)
    print("FILE NAME other", file_name)
    with open(base_image_url +"other/" + file_name, "rb") as image_contents:
        image_list.append(ImageFileCreateEntry(name=file_name, contents=image_contents.read(), tag_ids=[other.id]))


# Upload images with tags
upload_result = trainer.create_images_from_files(project.id, ImageFileCreateBatch(images=image_list))
if not upload_result.is_batch_successful:
    print("Image batch upload failed.")
    for image in upload_result.images:
        print("Image status: ", image.status, file_name)
    exit(-1)


# Start training
print ("Training...")
iteration = trainer.train_project(project.id)
while (iteration.status != "Completed"):
    iteration = trainer.get_iteration(project.id, iteration.id)
    print ("Training status: " + iteration.status)
    time.sleep(1)


# Publish the trainer
trainer.publish_iteration(project.id, iteration.id, publish_iteration_name, prediction_resource_id)
print ("Done!")

print("Project ID", project.id)
